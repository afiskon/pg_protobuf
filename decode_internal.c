#include "decode_ctx.h"
#include "decode_internal.h"
#include <postgres.h>
#include <port.h>

void protobuf_decode_type_and_tag(ProtobufDecodeCtx* ctx, uint8* type_res, uint32* tag_res);
uint64 protobuf_decode_value_or_length(ProtobufDecodeCtx* ctx);

/* Decode type and tag */
void protobuf_decode_type_and_tag(ProtobufDecodeCtx* ctx, uint8* type_res, uint32* tag_res) {
	uint8 cont = (*(ctx->protobuf_data)) >> 7;
	uint32 tag = ((*(ctx->protobuf_data)) >> 3) & 0x0F;
	uint8 type = (*(ctx->protobuf_data)) & 0x07;
	uint8 shift;

	protobuf_decode_ctx_shift(ctx, 1);

	shift = 4;
	while(cont) {
		if((shift / 8) >= sizeof(tag))
			ereport(ERROR,
				(
				 errcode(ERRCODE_INTERNAL_ERROR),
				 errmsg("Too long tag encoded in the protobuf data."),
				 errdetail("protobuf_decode_internal() - tag variable is "
							"uint32 and your protobuf stores larger tags."),
				 errhint("Sorry for that :( Patches are welcome!")
				));

		cont = (*(ctx->protobuf_data)) >> 7;
		tag = tag | (((uint32)((*(ctx->protobuf_data)) & 0x7F)) << shift);
		shift += 7;
		protobuf_decode_ctx_shift(ctx, 1);
	}

	*type_res = type;
	*tag_res = tag;
}

/* Decode value_or_length (int32, uint64, string length, etc) */
uint64 protobuf_decode_value_or_length(ProtobufDecodeCtx* ctx) {
	uint64 value_or_length = 0;
	uint8 shift = 0;
	uint8 temp;

	for(;;) {
		if((shift / 8) >= sizeof(value_or_length))
			ereport(ERROR,
				(
				 errcode(ERRCODE_INTERNAL_ERROR),
				 errmsg("Too long integer encoded in the protobuf data."),
				 errdetail("protobuf_decode_internal() - value_or_length variable is "
							"uint64 and your protobuf stores larger integers."),
				 errhint("That should have never happen. Your data is probably corrupted.")
				));

		temp = *ctx->protobuf_data;
		protobuf_decode_ctx_shift(ctx, 1);

		value_or_length = value_or_length | ( (((uint64)temp) & 0x7F) << shift);
		shift += 7;

		if(!(temp & 0x80)) // 0b0xxxxxxx - the end of the integer
			break;
	}

	return value_or_length;
}

/* Decode Protobuf structure. */
void protobuf_decode_internal(const uint8* protobuf_data, Size protobuf_size, ProtobufDecodeResult* result) {
	ProtobufDecodeCtx* ctx;
	ProtobufDecodeCtx decode_ctx;
	ctx = &decode_ctx;
	protobuf_decode_ctx_init(ctx, protobuf_data, protobuf_size);
	result->nfields = 0;

	while(ctx->protobuf_size > 0) {
		uint8 type;
		uint32 tag, offset = 0;
		uint64 value_or_length;

		if(result->nfields == PROTOBUF_RESULT_MAX_FIELDS)
			ereport(ERROR,
				(
				 errcode(ERRCODE_INTERNAL_ERROR),
				 errmsg("Protobuf data contains to many fields."),
				 errdetail("protobuf_decode_internal() - ProtobufDecodeResult structure can't fit all the fields."),
				 errhint("You can increase PROTOBUF_RESULT_MAX_FIELDS constant. "
						"Also it's probably worth notifying the author regarding this issue.")
				));

		/* Decode type and tag */
		protobuf_decode_type_and_tag(ctx, &type, &tag);

		switch(type) {
		case PROTOBUF_TYPE_INTEGER:
			/* Decode int32, uint64, etc */
			value_or_length = protobuf_decode_value_or_length(ctx);
			break;
		case PROTOBUF_TYPE_BYTES:
			/* Decode string length */
			value_or_length = protobuf_decode_value_or_length(ctx);
			/* Calculate string offset */
			offset = protobuf_decode_ctx_offset(ctx);
			/* Skip `length` bytes */
			protobuf_decode_ctx_shift(ctx, value_or_length);
			break;
		default:
			ereport(ERROR,
				(
				 errcode(ERRCODE_INTERNAL_ERROR),
				 errmsg("Usupported protobuf type."),
				 errdetail("pg_protobuf doesn't support type with id %u yet.", type),
				 errhint("Sorry for that :( Patches are welcome!")
				));
		}

		/* Save the field */
		result->field_info[result->nfields].tag = tag;
		result->field_info[result->nfields].type = type;
		result->field_info[result->nfields].value_or_length = (int64)value_or_length;
		result->field_info[result->nfields].offset = offset;
		result->nfields++;
	}
}

