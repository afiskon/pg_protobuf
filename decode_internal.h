/* vim: set ai et ts=4 sw=4: */

#ifndef _DECODE_INTERNAL_H_
#define _DECODE_INTERNAL_H_

#include <postgres.h>
#include <port.h>

#define PROTOBUF_TYPE_INTEGER 0
#define PROTOBUF_TYPE_FIXED64 1
#define PROTOBUF_TYPE_BYTES   2
#define PROTOBUF_TYPE_FIXED32 5

#define PROTOBUF_RESULT_MAX_FIELDS 256

typedef struct {
    uint32 tag;
    uint8 type;
    int64 value_or_length;
    uint32 offset; // for PROTOBUF_TYPE_BYTES only
} ProtobufFieldInfo;

typedef struct {
    uint32 nfields;
    ProtobufFieldInfo field_info[PROTOBUF_RESULT_MAX_FIELDS];
} ProtobufDecodeResult;

void protobuf_decode_internal(const uint8* protobuf_data, Size protobuf_size, ProtobufDecodeResult* result);

#endif // _DECODE_INTERNAL_H_ 
