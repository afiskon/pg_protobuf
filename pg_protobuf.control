comment = 'Protobuf support for PostgreSQL'
default_version = '1.0'
module_pathname = '$libdir/pg_protobuf'
relocatable = true
