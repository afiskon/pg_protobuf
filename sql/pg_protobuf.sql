CREATE EXTENSION pg_protobuf;

SELECT protobuf_decode(null);

SELECT protobuf_decode('garbage' :: bytea);

SELECT protobuf_decode(E'\\x0a0365617810321880022202100f' :: bytea);

SELECT protobuf_decode(E'\\x0a07616669736b6f6e10191880082a060a0200011064' :: bytea);

SELECT protobuf_get_integer(null, 1);

SELECT protobuf_get_integer(E'\\x0a07616669736b6f6e10191880082a060a0200011064' :: bytea, 1);

SELECT protobuf_get_integer(E'\\x0a07616669736b6f6e10191880082a060a0200011064' :: bytea, 2);

SELECT protobuf_get_integer(E'\\x0a07616669736b6f6e10191880082a060a0200011064' :: bytea, 3);

SELECT protobuf_get_bytes(null, 1);

SELECT protobuf_get_bytes(E'\\x0a0365617810321880022202100f' :: bytea, 1);

SELECT protobuf_get_bytes(E'\\x0a0365617810321880022202100f' :: bytea, 4);

SELECT protobuf_get_bytes(E'\\x0a0365617810321880022202100f' :: bytea, 5);

SELECT protobuf_get_bytes(E'\\x0a07616669736b6f6e10191880082a060a0200011064' :: bytea, 1);

SELECT protobuf_get_bytes(E'\\x0a07616669736b6f6e10191880082a060a0200011064' :: bytea, 4);

SELECT protobuf_get_bytes(E'\\x0a07616669736b6f6e10191880082a060a0200011064' :: bytea, 5);

SELECT protobuf_get_string(E'\\x0a0365617810321880022202100f' :: bytea, 1);

SELECT protobuf_get_string(E'\\x0a07616669736b6f6e10191880082a060a0200011064' :: bytea, 1);

SELECT protobuf_get_string(E'\\x10191880082a060a0200011064822007616669736b6f6e' :: bytea, 512);

SELECT protobuf_get_string(E'\\x10321880022202100f82200ad0a8d183d180d0b8d0ba' :: bytea, 512);

DROP EXTENSION pg_protobuf;
