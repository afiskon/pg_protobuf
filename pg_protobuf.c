/* vim: set ai et ts=4 sw=4: */

#include "decode_internal.h"
#include <postgres.h>
#include <port.h>
#include <catalog/pg_type.h>
#include <utils/builtins.h>
#include <utils/array.h>
#include <utils/lsyscache.h>

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(protobuf_decode);
PG_FUNCTION_INFO_V1(protobuf_get_int_multi);
PG_FUNCTION_INFO_V1(protobuf_get_sfixed32_multi);
PG_FUNCTION_INFO_V1(protobuf_get_float_multi);
PG_FUNCTION_INFO_V1(protobuf_get_sfixed64_multi);
PG_FUNCTION_INFO_V1(protobuf_get_double);
PG_FUNCTION_INFO_V1(protobuf_get_bytes);

typedef Datum (*FieldInfoToDatumCallback)(ProtobufFieldInfo*);

Datum protobuf_get_int_or_sfixed64_multi_callback(ProtobufFieldInfo* field_info);
Datum protobuf_get_sfixed32_multi_callback(ProtobufFieldInfo* field_info);
Datum protobuf_get_float_multi_callback(ProtobufFieldInfo* field_info);

Datum protobuf_get_any_multi_internal(
    bytea* protobuf_bytea, int32 tag, Oid element_type, uint8 expected_protobuf_type,
    FieldInfoToDatumCallback field_info_to_datum_cb);

/* protobuf -> cstring */
Datum protobuf_decode(PG_FUNCTION_ARGS) {
    char temp_buff[128];
    uint32 i;
    int len = -1;

    bytea* protobuf_bytea = PG_GETARG_BYTEA_P(0);
    Size protobuf_size = VARSIZE(protobuf_bytea) - VARHDRSZ;
    const uint8* protobuf_data = (const uint8*)VARDATA(protobuf_bytea);

    size_t result_buff_size = 1024;
    size_t result_buff_free = result_buff_size;
    char* result_buff = palloc(result_buff_size);

    ProtobufDecodeResult decode_result;
    protobuf_decode_internal(protobuf_data, protobuf_size, &decode_result);

    for(i = 0; i < decode_result.nfields; i++) {
        switch(decode_result.field_info[i].type) {
        case PROTOBUF_TYPE_INTEGER:
            len = snprintf(temp_buff, sizeof(temp_buff),
                "type = integer, tag = %u, value = %ld\n",
                decode_result.field_info[i].tag,
                decode_result.field_info[i].value_or_length);
            break;
        case PROTOBUF_TYPE_FIXED64:
            len = snprintf(temp_buff, sizeof(temp_buff),
                "type = fixed64, tag = %u, int value = %ld, float value = %.02f\n",
                decode_result.field_info[i].tag,
                decode_result.field_info[i].value_or_length,
                *(double*)&decode_result.field_info[i].value_or_length);
            break;

        case PROTOBUF_TYPE_BYTES:
            len = snprintf(temp_buff, sizeof(temp_buff),
                "type = bytes, tag = %u, length = %ld, offset = %u\n",
                decode_result.field_info[i].tag,
                decode_result.field_info[i].value_or_length,
                decode_result.field_info[i].offset);
            break;
        case PROTOBUF_TYPE_FIXED32:
            len = snprintf(temp_buff, sizeof(temp_buff),
                "type = fixed32, tag = %u, int value = %d, float value = %.02f\n",
                decode_result.field_info[i].tag,
                (int32)decode_result.field_info[i].value_or_length,
                *(float*)&decode_result.field_info[i].value_or_length);
            break;
        default:
            ereport(ERROR,
                (
                 errcode(ERRCODE_INTERNAL_ERROR),
                 errmsg("Unable to format protobuf data."),
                 errdetail("protobuf_decode() - unable to display type with id %u.", decode_result.field_info[i].type),
                 errhint("Most likely this is a bug in pg_protobuf. Please contact the author.")
                ));
        }

        if(len < 0)
            ereport(ERROR,
                (
                 errcode(ERRCODE_INTERNAL_ERROR),
                 errmsg("Unable to format protobuf data."),
                 errdetail("protobuf_decode() - snprintf() returned %d.", len),
                 errhint("Most likely this is a bug in pg_protobuf. Please contact the author.")
                ));

        if(len >= result_buff_free) {
            result_buff_free += result_buff_size;
            result_buff_size *= 2;
            repalloc(result_buff, result_buff_size);
        }

        memcpy(result_buff + (result_buff_size - result_buff_free), temp_buff, len+1);
        result_buff_free -= len;
    }

    PG_RETURN_CSTRING((Datum)result_buff);
}

/* Common code used in all protobuf_get_*_multi procedures */
Datum protobuf_get_any_multi_internal(
        bytea* protobuf_bytea, int32 tag, Oid element_type, uint8 expected_protobuf_type,
        FieldInfoToDatumCallback field_info_to_datum_cb) {
    Size protobuf_size = VARSIZE(protobuf_bytea) - VARHDRSZ;
    const uint8* protobuf_data = (const uint8*)VARDATA(protobuf_bytea);
    int  dims[MAXDIM], lbs[MAXDIM], nelements = 0, ndims = 1; /* one dimension */
    bool nulls[PROTOBUF_RESULT_MAX_FIELDS], typbyval;
    Datum elements[PROTOBUF_RESULT_MAX_FIELDS];
    int32 i;
    int16 typlen;
    char typalign;
    ProtobufDecodeResult decode_result;

    /* get required info about the array element type */
    get_typlenbyvalalign(element_type, &typlen, &typbyval, &typalign);

    /* decode the data */
    protobuf_decode_internal(protobuf_data, protobuf_size, &decode_result);

    /* fill the array */
    for(i = 0; i < decode_result.nfields; i++) {
        if(decode_result.field_info[i].tag != tag)
            continue;

        if(decode_result.field_info[i].type != expected_protobuf_type)
            /* if necessary, we can easily implement prtobuf_get_*_strict that whould throw an error */
            continue;

        elements[nelements] = field_info_to_datum_cb(&decode_result.field_info[i]);
        nulls[nelements] = false;
        nelements++;
    }

    /* build and return a pointer to ArrayType, PostgreSQL's array representation */
    dims[0] = nelements; /* number of elements */
    lbs[0] = 1; /* lower bound is 1 */
    PG_RETURN_POINTER(
        construct_md_array(elements, nulls, ndims, dims, lbs, element_type, typlen, typbyval, typalign)
    );
}

/* For internal usage in protobuf_get_{int,sfixed64}_multi procedures only */
Datum protobuf_get_int_or_sfixed64_multi_callback(ProtobufFieldInfo* field_info) {
    return Int64GetDatum(field_info->value_or_length);
}

/* protobuf -> int32[] / int64[] / etc */
Datum protobuf_get_int_multi(PG_FUNCTION_ARGS) {
    bytea* protobuf_bytea = PG_GETARG_BYTEA_P(0);
    int32 tag = PG_GETARG_INT32(1);

    return protobuf_get_any_multi_internal(
        protobuf_bytea, tag, INT8OID, PROTOBUF_TYPE_INTEGER,
        protobuf_get_int_or_sfixed64_multi_callback);
}

/* For internal usage in protobuf_get_sfixed32_multi procedure only */
Datum protobuf_get_sfixed32_multi_callback(ProtobufFieldInfo* field_info) {
    return Int32GetDatum(field_info->value_or_length);
}

/* protobuf -> sfixed32[] */
Datum protobuf_get_sfixed32_multi(PG_FUNCTION_ARGS) {
    bytea* protobuf_bytea = PG_GETARG_BYTEA_P(0);
    int32 tag = PG_GETARG_INT32(1);

    return protobuf_get_any_multi_internal(
        protobuf_bytea, tag, INT4OID, PROTOBUF_TYPE_FIXED32,
        protobuf_get_sfixed32_multi_callback);
}

/* For internal usage in protobuf_get_float_multi procedure only */
Datum protobuf_get_float_multi_callback(ProtobufFieldInfo* field_info) {
    return Float4GetDatum(*(float*)&(field_info->value_or_length));
}

/* protobuf -> float[] */
Datum protobuf_get_float_multi(PG_FUNCTION_ARGS) {
    bytea* protobuf_bytea = PG_GETARG_BYTEA_P(0);
    int32 tag = PG_GETARG_INT32(1);

    return protobuf_get_any_multi_internal(
        protobuf_bytea, tag, FLOAT4OID, PROTOBUF_TYPE_FIXED32,
        protobuf_get_float_multi_callback);
}

/* protobuf -> sfixed64[] */
Datum protobuf_get_sfixed64_multi(PG_FUNCTION_ARGS) {
    bytea* protobuf_bytea = PG_GETARG_BYTEA_P(0);
    int32 tag = PG_GETARG_INT32(1);

    return protobuf_get_any_multi_internal(
        protobuf_bytea, tag, INT8OID, PROTOBUF_TYPE_FIXED64,
        protobuf_get_int_or_sfixed64_multi_callback);
}

/* protobuf -> double */
Datum protobuf_get_double(PG_FUNCTION_ARGS) {
    bytea* protobuf_bytea = PG_GETARG_BYTEA_P(0);
    int32 i, tag = PG_GETARG_INT32(1);
    Size protobuf_size = VARSIZE(protobuf_bytea) - VARHDRSZ;
    const uint8* protobuf_data = (const uint8*)VARDATA(protobuf_bytea);
    ProtobufDecodeResult decode_result;

    protobuf_decode_internal(protobuf_data, protobuf_size, &decode_result);

    for(i = 0; i < decode_result.nfields; i++) {
        if(decode_result.field_info[i].tag == tag) {
            if(decode_result.field_info[i].type != PROTOBUF_TYPE_FIXED64)
                /* if necessary, we can easily implement prtobuf_get_*_strict that whould throw an error */
                PG_RETURN_NULL();

            PG_RETURN_FLOAT8(*(double*)&decode_result.field_info[i].value_or_length);
        }
    }

    /* not found */
    PG_RETURN_NULL();
}

/* protobuf -> bytea */
Datum protobuf_get_bytes(PG_FUNCTION_ARGS) {
    bytea* protobuf_bytea = PG_GETARG_BYTEA_P(0);
    int32 i, tag = PG_GETARG_INT32(1);
    Size protobuf_size = VARSIZE(protobuf_bytea) - VARHDRSZ;
    const uint8* protobuf_data = (const uint8*)VARDATA(protobuf_bytea);
    ProtobufDecodeResult decode_result;

    protobuf_decode_internal(protobuf_data, protobuf_size, &decode_result);

    for(i = 0; i < decode_result.nfields; i++) {
        if(decode_result.field_info[i].tag == tag) {
            if(decode_result.field_info[i].type != PROTOBUF_TYPE_BYTES)
                /* if necessary, we can easily implement prtobuf_get_*_strict that whould throw an error */
                PG_RETURN_NULL();
            else {
                uint64 length = (uint64)decode_result.field_info[i].value_or_length;
                uint8* buffer = palloc(VARHDRSZ + length);
                SET_VARSIZE(buffer, VARHDRSZ + length);
                memcpy(VARDATA(buffer), protobuf_data + decode_result.field_info[i].offset, length);
                PG_RETURN_BYTEA_P(buffer);
            }    
        }
    }    

    /* not found */
    PG_RETURN_NULL();
}

