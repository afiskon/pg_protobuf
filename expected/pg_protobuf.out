CREATE EXTENSION pg_protobuf;
SELECT protobuf_decode(null);
 protobuf_decode 
-----------------
 
(1 row)

SELECT protobuf_decode('garbage' :: bytea);
ERROR:  Usupported protobuf type.
DETAIL:  pg_protobuf doesn't support type with id 7 yet.
HINT:  Sorry for that :( Patches are welcome!
SELECT protobuf_decode(E'\\x0a0365617810321880022202100f' :: bytea);
                protobuf_decode                 
------------------------------------------------
 type = bytes, tag = 1, length = 3, offset = 2 +
 type = integer, tag = 2, value = 50           +
 type = integer, tag = 3, value = 256          +
 type = bytes, tag = 4, length = 2, offset = 12+
 
(1 row)

SELECT protobuf_decode(E'\\x0a07616669736b6f6e10191880082a060a0200011064' :: bytea);
                protobuf_decode                 
------------------------------------------------
 type = bytes, tag = 1, length = 7, offset = 2 +
 type = integer, tag = 2, value = 25           +
 type = integer, tag = 3, value = 1024         +
 type = bytes, tag = 5, length = 6, offset = 16+
 
(1 row)

SELECT protobuf_get_int(null, 1);
 protobuf_get_int 
------------------
                 
(1 row)

SELECT protobuf_get_int(E'\\x0a07616669736b6f6e10191880082a060a0200011064' :: bytea, 1);
 protobuf_get_int 
------------------
                 
(1 row)

SELECT protobuf_get_int(E'\\x0a07616669736b6f6e10191880082a060a0200011064' :: bytea, 2);
 protobuf_get_int 
------------------
               25
(1 row)

SELECT protobuf_get_int(E'\\x0a07616669736b6f6e10191880082a060a0200011064' :: bytea, 3);
 protobuf_get_int 
------------------
             1024
(1 row)

SELECT protobuf_get_bytes(null, 1);
 protobuf_get_bytes 
--------------------
 
(1 row)

SELECT protobuf_get_bytes(E'\\x0a0365617810321880022202100f' :: bytea, 1);
 protobuf_get_bytes 
--------------------
 \x656178
(1 row)

SELECT protobuf_get_bytes(E'\\x0a0365617810321880022202100f' :: bytea, 4);
 protobuf_get_bytes 
--------------------
 \x100f
(1 row)

SELECT protobuf_get_bytes(E'\\x0a0365617810321880022202100f' :: bytea, 5);
 protobuf_get_bytes 
--------------------
 
(1 row)

SELECT protobuf_get_bytes(E'\\x0a07616669736b6f6e10191880082a060a0200011064' :: bytea, 1);
 protobuf_get_bytes 
--------------------
 \x616669736b6f6e
(1 row)

SELECT protobuf_get_bytes(E'\\x0a07616669736b6f6e10191880082a060a0200011064' :: bytea, 4);
 protobuf_get_bytes 
--------------------
 
(1 row)

SELECT protobuf_get_bytes(E'\\x0a07616669736b6f6e10191880082a060a0200011064' :: bytea, 5);
 protobuf_get_bytes 
--------------------
 \x0a0200011064
(1 row)

SELECT protobuf_get_string(E'\\x0a0365617810321880022202100f' :: bytea, 1);
 protobuf_get_string 
---------------------
 eax
(1 row)

SELECT protobuf_get_string(E'\\x0a07616669736b6f6e10191880082a060a0200011064' :: bytea, 1);
 protobuf_get_string 
---------------------
 afiskon
(1 row)

SELECT protobuf_get_string(E'\\x10191880082a060a0200011064822007616669736b6f6e' :: bytea, 512);
 protobuf_get_string 
---------------------
 afiskon
(1 row)

SELECT protobuf_get_string(E'\\x10321880022202100ffaa3e80303656178', 999999);
 protobuf_get_string 
---------------------
 eax
(1 row)

SELECT protobuf_get_string(E'\\x10321880022202100f82200ad0a8d183d180d0b8d0ba' :: bytea, 512);
 protobuf_get_string 
---------------------
 Шурик
(1 row)

SELECT length(protobuf_get_string(E'\\x10321880022202100f82208e02' || convert_to(repeat('a', 270), 'utf-8'), 512));
 length 
--------
    270
(1 row)

SELECT protobuf_decode(E'\\x0a036561781088ef99abc5e88c91111880022202100f');
                   protobuf_decode                    
------------------------------------------------------
 type = bytes, tag = 1, length = 3, offset = 2       +
 type = integer, tag = 2, value = 1234605616436508552+
 type = integer, tag = 3, value = 256                +
 type = bytes, tag = 4, length = 2, offset = 20      +
 
(1 row)

SELECT protobuf_get_int(E'\\x0a036561781088ef99abc5e88c91111880022202100f', 2);
  protobuf_get_int   
---------------------
 1234605616436508552
(1 row)

SELECT protobuf_decode(E'\\x0a0365617810c0bbf8ffffffffffff0118f890e6d4ba97f3eeee012202100f');
                    protobuf_decode                    
-------------------------------------------------------
 type = bytes, tag = 1, length = 3, offset = 2        +
 type = integer, tag = 2, value = -123456             +
 type = integer, tag = 3, value = -1234605616436508552+
 type = bytes, tag = 4, length = 2, offset = 29       +
 
(1 row)

SELECT protobuf_get_int(E'\\x0a0365617810c0bbf8ffffffffffff0118f890e6d4ba97f3eeee012202100f', 2);
 protobuf_get_int 
------------------
          -123456
(1 row)

SELECT protobuf_get_int(E'\\x0a0365617810c0bbf8ffffffffffff0118f890e6d4ba97f3eeee012202100f', 3);
   protobuf_get_int   
----------------------
 -1234605616436508552
(1 row)

SELECT protobuf_get_int(E'\\x0a0365617810321880022202100f306f30de0130cd02', 6);
 protobuf_get_int 
------------------
              111
(1 row)

SELECT protobuf_get_int_multi(E'\\x0a0365617810321880022202100f306f30de0130cd02', 6);
 protobuf_get_int_multi 
------------------------
 {111,222,333}
(1 row)

SELECT protobuf_get_sfixed32(E'\\x0a0365617815320000001d00ffffff2202100f', 2);
 protobuf_get_sfixed32 
-----------------------
                    50
(1 row)

SELECT protobuf_get_sfixed32(E'\\x0a0365617815320000001d00ffffff2202100f', 3);
 protobuf_get_sfixed32 
-----------------------
                  -256
(1 row)

SELECT protobuf_get_float(E'\\x0a036561781500004a421d333380c32202100f', 2);
 protobuf_get_float 
--------------------
               50.5
(1 row)

SELECT protobuf_get_float(E'\\x0a036561781500004a421d333380c32202100f', 3);
 protobuf_get_float 
--------------------
             -256.4
(1 row)

SELECT protobuf_get_sfixed64(E'\\x0a036561781132000000000000001966666666660670c02202100f', 2);
 protobuf_get_sfixed64 
-----------------------
                    50
(1 row)

SELECT protobuf_get_sfixed64(E'\\x0a07616669736b6f6e11e7ffffffffffffff1933333333330390402a060a0200011064', 2);
 protobuf_get_sfixed64 
-----------------------
                   -25
(1 row)

SELECT protobuf_get_double(E'\\x0a036561781132000000000000001966666666660670c02202100f', 3);
 protobuf_get_double 
---------------------
              -256.4
(1 row)

SELECT protobuf_get_double(E'\\x0a07616669736b6f6e11e7ffffffffffffff1933333333330390402a060a0200011064', 3);
 protobuf_get_double 
---------------------
              1024.8
(1 row)

SELECT protobuf_get_bool(E'\\x0a03656178107b18c8032202100f', 6);
 protobuf_get_bool 
-------------------
 f
(1 row)

SELECT protobuf_get_bool(E'\\x0a07616669736b6f6e10191880082a060a02000110643001', 6);
 protobuf_get_bool 
-------------------
 t
(1 row)

DROP EXTENSION pg_protobuf;
